New module: Add ec2 Vpn Connection module and tests (cloud/amazon/ec2_vpc_vpn)

s-hertel opened the pull request:
##### SUMMARY
Adding a VPN Connection module. Can create/tag/modify tags/delete vpc vpn connections. Unit tests use placebo fixtures. 

Edit: To run the tests: since placebo's default region is us-west-2 be sure your default region is set to that and that the environment variable PLACEBO_RECORD is false or unset. I set the environment var AWS_DEFAULT_PROFILE and run with `$ pytest test/units/modules/cloud/amazon/test_ec2_vpc_vpn.py`.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py
test/units/modules/cloud/amazon/test_ec2_vpc_vpn.py

##### ANSIBLE VERSION
```
2.4.0
```

s-hertel committed:
Add VPN module and unittests

s-hertel committed:
remove unnecessary imports

s-hertel committed:
fix documentation

gundalow commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,499 @@
+DOCUMENTATION = """
+---
+module: ec2_vpc_vpn
+short_description: create, modify, and delete VPN connections
```
`Create, modify, and delete EC2 VPN connections`

gundalow commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,499 @@
+options:
+  state:
+    description:
+      - the desired state of the vpn connection.
```
The

gundalow commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,499 @@
+    module = AnsibleModule(argument_spec=argument_spec)
+
+    if not HAS_BOTO3:
+        module.fail_json(msg='boto required for this module')
```
boto3?

ryansb commented on test/units/modules/cloud/amazon/test_ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,336 @@
+import pytest
+import os
+from . placebo_fixtures import placeboify, maybe_sleep
+from ansible.modules.cloud.amazon import ec2_vpc_vpn as v
```
Can you use the full name here? Single-letter vars aren't great for readability.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,499 @@
+    default: False
+    required: no
+  filters:
+    description:
```
This means any sufficiently unique set of filters, right? So I could find one based on just the BGP ASN and a tag, and that would work so long as it was unique?

s-hertel replied:
Yes, that's right. I am going to document this a little better with `suboptions:` and I'm still debating the filter names because `route.destination-cidr-block` doesn't feel very user friendly. I'll probably rename some things. Also need to verify if these are the correct types for each option and add a few examples and another unittest or two.

s-hertel committed:
raise custom exception rather than passing module everywhere

s-hertel committed:
remove recordings

Rerecord tests

Fix docs

Ensure vpn_connection_id is a list of strings when checking if it exists

s-hertel committed:
fix check mode

s-hertel committed:
Rerecord tests

s-hertel committed:
remove superfluous code and comments and make exception handling uniform

fix docs

s-hertel committed:
Fix ec2_vpc_vpn documentation

s-hertel committed:
make ec2_vpc_vpn compatible with python 2.6 and make check mode logic more succinct

s-hertel committed:
fix comparison of list of dicts

willthames commented:
Do we really need 113 new files as a result of one new module?

bcoca commented:
@willthames 112 are just for tests

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,745 @@
+short_description: Create, modify, and delete EC2 VPN connections.
+description:
+  - This module creates, modifies, and deletes VPN connections. Idempotence is achieved by using the filters
+    options or specifying the VPN connection identifier.
```
s/options/option/

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,745 @@
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_text
+from ansible.module_utils.ec2 import (boto3_conn, get_aws_connection_info, ec2_argument_spec,
```
If you're using this many things from module_utils.ec2, I'd just do `from ansible.module_utils import ec2 as ec2_utils` or similar.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,745 @@
+def add_routes(connection, vpn_connection_id, routes_to_add):
+    try:
+        for route in routes_to_add:
+            response = connection.create_vpn_connection_route(VpnConnectionId=vpn_connection_id,
```
The `response` here is never used.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,745 @@
+def remove_routes(connection, vpn_connection_id, routes_to_remove):
+    try:
+        for route in routes_to_remove:
+            response = connection.delete_vpn_connection_route(VpnConnectionId=vpn_connection_id,
```
Unused `response`

ryansb replied:
Also, you should probably swap the order of the try and for, so you can add which route was being added when the module failed. This might be helpful in finding malformed IPs or other weird stuff.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,745 @@
+
+
+def create_filter(module_params, provided_filters):
+    ''' Creates a filter using the user-specified parameters and unmodifiable options that may have been specified in the task '''
```
`s/'''/"""/`

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,745 @@
+        # fix filter names to be recognized by boto3
+        if raw_param in boto3ify_filter:
+            param = boto3ify_filter[raw_param]
+            provided_filters[param] = provided_filters[raw_param]
```
You can do in-place renames like:

```
provided_filters[param] = provided_filters.pop(raw_param)
```

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,745 @@
+            flat_filter_dict[param_to_filter[param]] = [module_params.get(param)]
+
+    # change the flat dict into something boto3 will understand
+    formatted_filter = [{'Name': key, 'Values': flat_filter_dict[key]} for key in flat_filter_dict]
```
I think `[{'Name': key, 'Values': value} for key, value in flat_filter_dict.items()]` would be a clearer way of doing this.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,745 @@
+        if connections['VpnConnections'][0]['State'] not in ("deleted", "deleting"):
+            return connections['VpnConnections'][0]
+        # Found the result but it was deleted already; since there was only one result create a new one
+        else:
```
This `else` and the one on line 436 can both be removed since the default return of `None` will be equivalent.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,745 @@
+
+
+def delete_connection(connection, vpn_connection_id):
+    ''' Deletes a VPN connection '''
```
s/'''/"""/

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,745 @@
+
+
+def check_for_update(connection, module_params, vpn_connection_id):
+    ''' Determines if there are any tags or routes that need to be updated. Ensures non-modifiable attributes aren't expected to change. '''
```
s/'''/"""/

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py:
```diff
@@ -0,0 +1,745 @@
+
+
+def make_changes(connection, vpn_connection_id, changes):
+    '''
```
s/'''/"""/

s-hertel committed:
Fix typos and docstrings

make requested changes for imports

make code clearer

s-hertel commented:
ready_for_review

ryansb commented:
shipit

ryansb commented:
rebuild_merge

s-hertel committed:
Fix copyright and metadata version.

s-hertel commented:
rebuild_merge
